### PR TITLE
Fix graph failed to register config to meta.

### LINKED ISSUE #780: [2 bugs] about show variables
1. the service has started for about an hour, but why `show variables graph` is empty
2. why `show variables graph` two times return different?  -- I can't reproduce the syntax error

image
image

grep don't return any `update config in cache `

image

### PR DESCRIPTION
**Pull request by critical27**
1. Let meta client retry to register configs to meta server.
2. Add `gflags.json` to avoid hard code config mode, and all default config mode is `IMMUTABLE`.
3. Add related command in complete.json

close #780

### CODE REVIEW

File: src/meta/client/MetaClient.cpp [hunk 1]
```diff
@@ -1055,6 +1056,9 @@ MetaClient::regConfig(const std::vector<cpp2::ConfigItem>& items) {
 
 folly::Future<StatusOr<std::vector<cpp2::ConfigItem>>>
 MetaClient::getConfig(const cpp2::ConfigModule& module, const std::string& name) {
+    if (!configReady_) {
+        registerCfg();
```
Comment (laura-ding) [inline src/meta/client/MetaClient.cpp]: getConfig() used in graphd's executor, and registerCfg() is a sync call, in graphd executor should not exist sync call. Can we do` if configReady_ is not true, direct return error, like the schema?`

File: src/meta/GflagsManager.cpp [hunk 2]
```diff
@@ -23,19 +25,59 @@ std::string GflagsManager::gflagsValueToThriftValue<std::string>(
...
 
-void GflagsManager::declareGflags() {
-    // declare all gflags in ClientBasedGflagsManager
+std::unordered_map<std::string, cpp2::ConfigMode> GflagsManager::parseConfigJson() {
+    auto path = fs::FileUtils::readLink("/proc/self/exe").value();
+    auto json = fs::FileUtils::dirname(path.c_str()) + "/../share/resources/gflags.json";
+    std::unordered_map<std::string, cpp2::ConfigMode> configModeMap;
+    Configuration conf;
+    if (!conf.parseFromFile(json).ok()) {
+        return configModeMap;
```
Comment (laura-ding) [inline src/meta/GflagsManager.cpp]: add error log

File: src/meta/client/MetaClient.cpp (hunk 1 above)
Comment (critical27) [reply src/meta/client/MetaClient.cpp]: Yeah, it do block graphd. The reason I add this logic is that if graphd register failed during initialization, that would take another `load_data_interval_secs`, during that time users can't show/get/update, do you think it's ok?

Comment (laura-ding) [reply src/meta/client/MetaClient.cpp]: yes

**Comment by critical27** [comment]
Jenkins go

File: src/meta/GflagsManager.cpp [hunk 3]
```diff
@@ -23,19 +25,60 @@ std::string GflagsManager::gflagsValueToThriftValue<std::string>(
...
+std::unordered_map<std::string, cpp2::ConfigMode> GflagsManager::parseConfigJson() {
+    auto path = fs::FileUtils::readLink("/proc/self/exe").value();
+    auto json = fs::FileUtils::dirname(path.c_str()) + "/../share/resources/gflags.json";
+    std::unordered_map<std::string, cpp2::ConfigMode> configModeMap;
+    Configuration conf;
+    if (!conf.parseFromFile(json).ok()) {
+        LOG(ERROR) << "Load gflags json failed";
+        return configModeMap;
+    }
+    std::vector<std::string> keys = {"IMMUTABLE", "REBOOT", "MUTABLE", "IGNORED"};
```
Comment (wadeliuyi) [inline src/meta/GflagsManager.cpp]: static local variable?

File: src/meta/GflagsManager.cpp [hunk 4]
```diff
@@ -23,19 +25,60 @@ std::string GflagsManager::gflagsValueToThriftValue<std::string>(
...
-        cpp2::ConfigMode mode = cpp2::ConfigMode::REBOOT;
         VariantType value;
         std::string valueStr;
 
+        // default config type would be immutable
+        cpp2::ConfigMode mode = cpp2::ConfigMode::IMMUTABLE;
+        if (configModeMap.find(name) != configModeMap.end()) {
+            mode = configModeMap[name];
+        }
+        // ignore some useless gflags
```
Comment (wadeliuyi) [inline src/meta/GflagsManager.cpp]: find and [] will traverse the map twice.

auto iter = configModeMap.find(name);
if (iter != configModeMap.end()) {
    mode = iter->second;
}

File: src/meta/GflagsManager.cpp [hunk 5]
```diff
@@ -23,19 +25,60 @@ std::string GflagsManager::gflagsValueToThriftValue<std::string>(
...
+        if (mode == cpp2::ConfigMode::IGNORED) {
```
Comment (wadeliuyi) [inline src/meta/GflagsManager.cpp]: mode default is cpp2::ConfigMode::IMMUTABLE, if configModeMap.find(name) != configModeMap.end() is false, this condition never true, is it right to go the next logic?

File: src/meta/GflagsManager.cpp (hunk 4 above)
Comment (critical27) [reply src/meta/GflagsManager.cpp]: nice catch.

File: src/meta/GflagsManager.cpp (hunk 3 above)
Comment (critical27) [reply src/meta/GflagsManager.cpp]: good suggestion

File: src/meta/GflagsManager.cpp (hunk 5 above)
Comment (critical27) [reply src/meta/GflagsManager.cpp]: The configModeMap is load from json, if found in json, replace the default mode. So for those configs we want to ignore, the mode will set to  cpp2::ConfigMode::IGNORED, and we can just ignore those gflags.

File: share/resources/gflags.json [hunk 6]
```diff
@@ -0,0 +1,37 @@
+{
```
Comment (laura-ding) [inline share/resources/gflags.json]: add gflags.json into share/resources/CMakeLists.txt

**Comment by critical27** [comment]
Jenkins go

File: src/meta/GflagsManager.cpp [hunk 7]
```diff
@@ -23,19 +25,63 @@ std::string GflagsManager::gflagsValueToThriftValue<std::string>(
...
+            continue;
+        }
+        if (module == cpp2::ConfigModule::META) {
+            // all config of meta is immutable for now
+            mode = cpp2::ConfigMode::IMMUTABLE;
+        }
+
         // TODO: all int32 and uint32 are converted to int64
         if (type == "uint32" || type == "int32" || type == "int64") {
```
Comment (dangleptr) [inline src/meta/GflagsManager.cpp]: Add uint64

File: share/resources/gflags.json (hunk 6 above)
Comment (dangleptr) [inline share/resources/gflags.json]: That's great.  One question,  how to distinguish different module,  storage/graph/meta  has different gflags.

File: src/meta/GflagsManager.cpp (hunk 7 above)
Comment (critical27) [reply src/meta/GflagsManager.cpp]: Nice catch.

File: share/resources/gflags.json (hunk 6 above)
Comment (critical27) [reply share/resources/gflags.json]: For now it seems we can apply theres to all storage/graph/meta, if we do need to distinguish modules of a gflag, I will add another layer to indicate modules.

Comment (dangleptr) [reply share/resources/gflags.json]: Yes, we need it.   Maybe we will start graph/meta/storage on the same server

File: share/resources/gflags.json [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    "IMMUTABLE": [
```
Comment (dangleptr) [inline share/resources/gflags.json]: Could we start all services on one host with just one gflags.json?

Comment (critical27) [reply share/resources/gflags.json]: Yeah, all services will try to read this file.

**Pull request merged**